File: frontend/src/lib/bolusTrace.js
function finiteOrNull(value) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

/**
 * Build a retrospective trace from an already-computed backend response.
 * This object is persisted for audit/learning only and is never a dosing input.
 */
export function buildClientBolusTrace(result, acceptedU, source = 'app') {
  const calc = result?.calc || result || {};
  const glucose = calc?.glucose || null;
  const used = calc?.used_params || calc?.usedParams || {};

  const recommendedU = finiteOrNull(calc?.total_u_final ?? calc?.total_u);
  const accepted = finiteOrNull(acceptedU);

  return {
    snapshot: {
      schema_version: 1,
      source,
      recommended_u: recommendedU,
      accepted_u: accepted,
      kind: result?.kind || calc?.kind || 'normal',
      meal_component_u: finiteOrNull(calc?.meal_bolus_u),
      correction_component_u: finiteOrNull(calc?.correction_u),
      iob_u: finiteOrNull(calc?.iob_u),
      iob_applied_to_correction_u: finiteOrNull(calc?.iob_applied_to_correction_u),
      total_u_raw: finiteOrNull(calc?.total_u_raw),
      total_u_final: recommendedU,
      upfront_u: finiteOrNull(result?.upfront_u ?? calc?.upfront_u),
      later_u: finiteOrNull(result?.later_u ?? calc?.later_u),
      duration_min: finiteOrNull(result?.duration_min ?? calc?.duration_min),
      glucose,
      warnings: Array.isArray(calc?.warnings) ? [...calc.warnings] : [],
      assumptions: Array.isArray(calc?.assumptions) ? [...calc.assumptions] : [],
      explain: Array.isArray(calc?.explain) ? [...calc.explain] : [],
    },
    applied_ratios: { ...used },
    context: {
      bg: finiteOrNull(glucose?.mgdl),
      trend: glucose?.trend ?? null,
      iob: finiteOrNull(calc?.iob_u),
      glucose_source: glucose?.source ?? null,
      glucose_age_minutes: finiteOrNull(glucose?.age_minutes),
      glucose_is_stale: Boolean(glucose?.is_stale),
    },
  };
}

File: backend/app/services/bolus_trace.py
from __future__ import annotations

from typing import Any, Optional

from app.models.bolus_v2 import BolusResponseV2

TRACE_SCHEMA_VERSION = 1


def build_bolus_trace(
    rec: BolusResponseV2,
    *,
    accepted_u: Optional[float] = None,
    source: Optional[str] = None,
) -> tuple[dict[str, Any], dict[str, Any], dict[str, Any]]:
    """Build a retrospective, non-dosing snapshot of one recommendation.

    The returned dictionaries intentionally contain no credentials, Nightscout
    tokens or other secrets. They exist only to explain/replay what the engine
    used at recommendation time; they must never be treated as future dosing
    inputs.
    """
    glucose = rec.glucose
    used = rec.used_params

    snapshot = {
        "schema_version": TRACE_SCHEMA_VERSION,
        "source": source,
        "recommended_u": rec.total_u_final,
        "accepted_u": accepted_u,
        "kind": rec.kind,
        "meal_component_u": rec.meal_bolus_u,
        "correction_component_u": rec.correction_u,
        "iob_u": rec.iob_u,
        "total_u_raw": rec.total_u_raw,
        "total_u_final": rec.total_u_final,
        "upfront_u": rec.upfront_u,
        "later_u": rec.later_u,
        "duration_min": rec.duration_min,
        "glucose": glucose.model_dump(mode="json") if glucose else None,
        "warnings": list(rec.warnings or []),
        "assumptions": list(rec.assumptions or []),
        "explain": list(rec.explain or []),
    }

    applied_ratios = used.model_dump(mode="json") if used else {}

    context = {
        "bg": glucose.mgdl if glucose else None,
        "trend": glucose.trend if glucose else None,
        "iob": rec.iob_u,
        "glucose_source": glucose.source if glucose else None,
        "glucose_age_minutes": glucose.age_minutes if glucose else None,
        "glucose_is_stale": glucose.is_stale if glucose else None,
    }
    return snapshot, applied_ratios, context
